src/aspire/noise/noise.py

```diff
@@ -345,3 +354,297 @@ def estimate_noise_psd(self):
+
+        Port of MATLAB `cryo_epsdS`.
+
+        :param images: Images instance
+        :param samples_idx: Boolean mask shaped (L,L).
+        :param max_d: Max computed correlation distance in pixels.
+           Default of `None` yields `np.floor(L / 3)`.
+        :param batch_size:  The size of the batches in which to compute the variance estimate.
+        :return:
+            - 2D PSD array
+            - Radial PSD array
+            - Distances map array
+            - Count of nonzero correlations array
+        """
+
+        n_img = images.n_images
+        L = samples_idx.shape[-1]
+        batch_size = min(batch_size, n_img)
+
+        # Migrate mask to GPU as needed
+        _samples_idx = xp.asarray(samples_idx)
+
+        # Correlations more than `max_d` pixels apart are not computed.
+        if max_d is None:
+            max_d = np.floor(L / 3)
+        if max_d > L - 1:
+            logger.info(
+                f"`max_d` value {max_d}greater than number of image pixels {L}, clipping to {L-1}."
+            )
+        max_d = int(min(max_d, L - 1))
+
+        R, x, _ = LegacyNoiseEstimator.estimate_power_spectrum_distribution_1d(
+            images=images, samples_idx=samples_idx, max_d=max_d, batch_size=batch_size
+        )
+        _R = xp.asarray(R)  # Migrate to GPU for assignments below
+
+        # Use the 1D autocorrelation estimated above to populate an
+        # array of the 2D isotropic autocorrelction. This
+        # autocorrelation is later Fourier transformed to get the
+        # power spectrum.
+        R2 = xp.zeros((2 * L - 1, 2 * L - 1), dtype=np.float64)
+
+        X, Y = xp.mgrid[-L + 1 : L, -L + 1 : L]
+        dists2 = X * X + Y * Y
+        dsquare2 = xp.sort(xp.unique(dists2[dists2 <= max_d**2]))
+        for i, d in enumerate(dsquare2):
+            idx = dists2 == d
+            R2[idx] = _R[i]
+
+        # Window the 2D autocorrelation and Fourier transform it to get the power
+        # spectrum. Always use the Gaussian window, as it has positive Fourier
+        # transform.
+        w = xp.asarray(gaussian_window(L, max_d))
+        P2 = fft.centered_fft2(R2 * w)
+        if (err := xp.linalg.norm(P2.imag) / xp.linalg.norm(P2)) > 1e-12:
+            logger.warning(f"Large imaginary components in P2 {err}.")
+        P2 = P2.real
+
+        # Normalize the power spectrum P2. The power spectrum is normalized such
+        # that its energy is equal to the average energy of the noise samples used
+        # to estimate it.
+
+        E = 0.0  # Total energy of the noise samples used to estimate the power spectrum.
+        samples = xp.zeros((batch_size, L, L), dtype=np.float64)
+        for start in trange(0, n_img, batch_size, desc="Estimating image noise energy"):
+            end = min(n_img, start + batch_size)
+            cnt = end - start
+
+            samples[:cnt, _samples_idx] = images[start:end].asnumpy()[0][samples_idx]
+            E += xp.sum(
```

Comment:
Are we sure this is a good idea? Makes `E` dependent on `batch_size`. Other option would be to track first and second moments, then compute variance afterwards.

Reply:
Unless I have a bug that is not what this should be doing. 

If I read correctly the original code was computing a per image mean. ie, demeaning per image. 

My code should be computing `cnt` different means, and subracting each image's mean.  Axis (1,2) should collapse the image axis, leaving the `cnt` stack axis.  The reshape is there to be explicit about the broadcast.

Reply:
Sorry my bad. I read this too quickly.